File: 1632.rank-transform-of-a-matrix.py
#
# @lc app=leetcode id=1632 lang=python3
#
# [1632] Rank Transform of a Matrix
#

# @lc code=start
from typing import List

class UnionFindSet:
    def __init__(self, n=0):
        self.parents = [i for i in range(n)]
        self.count = n

    def find(self, u):
        if u != self.parents[u]:
            self.parents[u] = self.find(self.parents[u])
        return self.parents[u]

    def union(self, u, v):
        pu, pv = self.find(u), self.find(v)
        if pu != pv:
            pMax, pMin = max(pu,pv), min(pu,pv)
            self.parents[pMax] = pMin
            self.count -= 1

class Solution:
    # We first sort the values in matrix
    # Then we build the graph from small value to large
    # The trick is for a value appearing multiple times in the matrix, 
    # we need to build the disjoint-set structure
    # This structure has m + n elements (m = row number, n = column number)
    # For each element belonging to the value, we union up the corresponding row and col.
    # Then we loop through the rows and columns and assign rankMax[find[i]] = max(rank[find[i]], rank[i])
    # We also record the inverse mapping from find[i] to each i
    # Therefore the find[i] for each cluster should contain the maximum rank for the row-group in each round
    # So update the new rank and assign the result
    def matrixRankTransform(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        hashTable = {}
        for i in range(m):
            for j in range(n):
                if matrix[i][j] not in hashTable:
                    hashTable[matrix[i][j]] = []
                hashTable[matrix[i][j]].append([i,j])

        result = [[0 for j in range(n)] for i in range(m)]
        rank = [0]* (m + n)
        for value in sorted(hashTable.keys()):
            UF = UnionFindSet(m + n)
            changed = set()
            for x, y in hashTable[value]:
                UF.union(x, y + m)
                changed.add(x)
                changed.add(y + m)
            rankMax = rank.copy()
            for i in changed:
                j = UF.find(i)
                rankMax[j] = max(rankMax[j], rank[i])
            for i in changed:
                rank[i] = rankMax[UF.find(i)] + 1
            for x, y in hashTable[value]:
                result[x][y] = rank[x]
        return result            
# @lc code=end

File: 1632.rank-transform-of-a-matrix.cs
/*
 * @lc app=leetcode id=1632 lang=csharp
 *
 * [1632] Rank Transform of a Matrix
 */

// @lc code=start
public class UnionFindSet
{
    public int[] parents{get; set;}
    public int count{get; set;} 
    public UnionFindSet(int n = 0)
    {
        parents = Enumerable.Range(0, n).ToArray();
        count = n;
    }

    public int find(int u)
    {
        if (u != parents[u])
        {
            parents[u] = find(parents[u]);
        }
        return parents[u];
    }

    public void union(int u, int v)
    {
        int pu = find(u);
        int pv = find(v);
        if (pu != pv)
        {
            int pMax = Math.Max(pu, pv);
            int pMin = Math.Min(pu, pv);
            parents[pMax] = pMin;
            count--;
        }
    }
}
public class Solution 
{
    public int[][] MatrixRankTransform(int[][] matrix) 
    {
        int m = matrix.Length;
        int n = matrix[0].Length;
        SortedDictionary<int, List<int[]>> hashTable = new SortedDictionary<int, List<int[]>>();
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (!hashTable.ContainsKey(matrix[i][j]))
                {
                    hashTable.Add(matrix[i][j], new List<int[]>());
                }
                hashTable[matrix[i][j]].Add(new int[2]{i, j});
            }
        }

        int[][] result = new int[m][];
        for (int i = 0; i < m; i++)
        {
            result[i] = new int[n];
            Array.Fill(result[i], 0);
        }
        int[] rank = new int[m + n];
        Array.Fill(rank, 0);
        foreach (int value in hashTable.Keys)
        {
            UnionFindSet UF = new UnionFindSet(m + n);
            HashSet<int> changed = new HashSet<int>();
            foreach (int[] item in hashTable[value])
            {
                UF.union(item[0], item[1] + m);
                changed.Add(item[0]);
                changed.Add(item[1] + m);
            }
            int[] rankMax = new int[m + n];
            Array.Copy(rank, rankMax, m + n);
            foreach (int i in changed)
            {
                int j = UF.find(i);
                rankMax[j] = Math.Max(rankMax[j], rank[i]);
            }
            foreach (int i in changed)
            {
                rank[i] = rankMax[UF.find(i)] + 1;
            }           
            foreach (int[] item in hashTable[value])
            {
                result[item[0]][item[1]] = rank[item[0]];
            }
        }
        return result;
    }
}
// @lc code=end

